src/hyperparams.rs

```diff
@@ -5,22 +5,19 @@ use crate::{
     traits::{Fit, IncrementalFit, Transformer},
 };
 
-/// A set of hyperparameters whose values have not been checked for validity. A reference to the
+/// A set of parameters whose values have not been checked for validity. A reference to the
 /// checked hyperparameters can only be obtained after checking has completed. If the
 /// `Transformer`, `Fit`, or `IncrementalFit` traits have been implemented on the checked
 /// hyperparameters, they will also be implemented on the unchecked hyperparameters with the
 /// checking step done automatically.
 ///
 /// The hyperparameter validation done in `check_ref()` and `check()` should be identical.
-pub trait UncheckedHyperParams {
-    /// The checked hyperparameters
+pub trait ParameterCheck {
     type Checked;
-    /// Error type resulting from failed hyperparameter checking
     type Error: Error;
 
-    /// Checks the hyperparameters and returns a reference to the checked hyperparameters if
-    /// successful
-    fn check_ref(&self) -> Result<&Self::Checked, Self::Error>;
```

Comment:
Removing `check_ref` is iffy because the blanket impls for `Predict`, `Fit`, and `IncrementalFit` rely on it. Without the blanket impls you'd need to have "checked" and "unchecked" impls of those 3 traits for every relevant parameter type, which is non-ideal. Admittedly `check_ref` doesn't work with the const generic guard pattern without some crazy reference casting, but I believe the ergonomics of const generic guards don't outweigh the benefits of having `check_ref` blanket impls.

Reply:
[Here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=32e4579ff04cdfa78b8c5d685d6f5a51) is the reference cast required to make `check_ref` work with const generics. It's unsafe but probably not UB as long as the underlying layout is identical. At least Miri is fine with it.

Reply:
thanks I was looking for something like this